src/mon/OSDMonitor.cc

```diff
@@ -7784,22 +7784,54 @@ int OSDMonitor::get_crush_rule(const string &rule_name,
   return 0;
 }
 
-int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, ostream *ss)
+int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, int crush_rule, ostream *ss)
 {
   auto max_pgs_per_osd = g_conf().get_val<uint64_t>("mon_max_pg_per_osd");
-  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u);   // assume min cluster size 3
-  auto max_pgs = max_pgs_per_osd * num_osds;
   uint64_t projected = 0;
-  if (pool < 0) {
+  unsigned osd_num = 0;
+  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u); // assume min cluster size 3
+  CrushWrapper newcrush = _get_pending_crush();
+  if (mapping.get_epoch() >= osdmap.get_epoch()) {
+    set<int> roots;
+    newcrush.find_takes_by_rule(crush_rule, &roots);
+    int max_osd = osdmap.get_max_osd();
+    for (auto root : roots) {
+      const char *rootname = newcrush.get_item_name(root);
+      set<int> osd_ids;
+      newcrush.get_leaves(rootname, &osd_ids);
+      unsigned out_osd = 0;
+      for (auto id : osd_ids) {
+	if (id >= max_osd) { 
+	  out_osd++;
+	  continue;
+	}
+	projected += mapping.get_osd_acting_pgs(id).size();
+      }
+      osd_num += osd_ids.size() - out_osd;
+    }
+    if (pool >= 0) {
+      for (const auto& i : osdmap.get_pools()) {
+	if (i.first == pool) {
+	  projected -= i.second.get_pg_num_target() * i.second.get_size();
+	  break;
+	}
+      }
+    }
     projected += pg_num * size;
-  }
-  for (const auto& i : osdmap.get_pools()) {
-    if (i.first == pool) {
+    num_osds = std::max(osd_num, 3u);  // assume min cluster size 3
+  } else {
+    if (pool < 0) {
       projected += pg_num * size;
-    } else {
-      projected += i.second.get_pg_num_target() * i.second.get_size();
+    }
+    for (const auto& i : osdmap.get_pools()) {
```

Comment:
since you are at here, might want to use the structured binding to refactor this loop, like:
```c++
for (const auto&[pool_id, pool_info] : osdmap.get_pools()) {
  if (pool_id != pool) {
    projected += pool_info.get_pg_num_target() * pool_info.get_size();
  }
}
```